### PR TITLE
Implement Kotlin Coroutines Plugin

### PR DESCRIPTION
**Pull request by bastrich**
Implemented #1963 

Here I build CoroutineInterceptor and CoroutineWrapper classes manually with ByteBuddy and in the local scope of the advice method. Otherwise, if I build them in another static context or just write like simple Java/Kotlin classes, I get ClassDefNotFoundError for Kotlin classes when integrating with traced applications. I don't know why, but building classes with ByteBuddy in the local scope of advice function resolved that issue. And I wrote an integration test for such case.

- [x] This is a new plugin
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [x] My code follows the [style guidelines of this project](https://github.com/elastic/apm-agent-java/blob/master/CONTRIBUTING.md#java-language-formatting-guidelines)
  - [x] I have made corresponding changes to the documentation
  - [x] I have added tests that prove my fix is effective or that my feature works
  - [x] New and existing [**unit** tests](https://github.com/elastic/apm-agent-java/blob/master/CONTRIBUTING.md#testing) pass locally with my changes
  - [x] I have updated [supported-technologies.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/docs/supported-technologies.asciidoc)
  - [x] Added an API method or config option? Document in which version this will be introduced
  - [x] Added an instrumentation plugin? Describe how you made sure that old, non-supported versions are not instrumented by accident.

### CODE REVIEW

**Comment by bastrich** [comment]
Do I need to change project.version manually or here is automatic version bump?

**Comment by bastrich** [comment]
@felixbarny should I add you to this PR? or how can it be reviewed?

**Comment by thecloudgeek** [comment]
can we get this one implemented soon?

**Comment by bastrich** [comment]
> can we get this one implemented soon?

not sure, see this thread on the forum - https://discuss.elastic.co/t/support-kotlin-coroutines/281846

**Comment by patiramyadav** [comment]
> > can we get this one implemented soon?
> 
> @thecloudgeek not sure, see this thread on the forum - https://discuss.elastic.co/t/support-kotlin-coroutines/281846
> 
> but generally it works, you can copy the source code and build your own plugin

@bastrich How do I use this plugin in kotlin gradle based project? I have cloned this project and built locally. I have plugin jar file now.

**Comment by patiramyadav** [comment]
```
2021-09-29 22:35:46,770 [reactor-http-nio-3] DEBUG co.elastic.apm.agent.mdc.MdcActivationListener - Class loader jdk.internal.loader.ClassLoaders$AppClassLoader@9e89d68 cannot load log4j API
java.lang.ClassNotFoundException: org.apache.log4j.MDC
	at jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581) ~[?:?]
	at jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:522) ~[?:?]
	at co.elastic.apm.agent.mdc.MdcActivationListener$2.get(MdcActivationListener.java:74) ~[elastic-apm-agent-1.26.1-SNAPSHOT.jar:?]
	at co.elastic.apm.agent.mdc.MdcActivationListener$2.get(MdcActivationListener.java:68) ~[elastic-apm-agent-1.26.1-SNAPSHOT.jar:?]
	at co.elastic.apm.agent.cache.WeakKeySoftValueLoadingCache$CacheValue.<init>(WeakKeySoftValueLoadingCache.java:119) ~[elastic-apm-agent-1.
```
I have 
```
    // logging deps
    implementation 'io.github.microutils:kotlin-logging:1.7.2'
```

File: apm-agent-plugins/apm-kotlin-concurrent-plugin/src/main/java/co/elastic/apm/agent/kotlinconcurrent/CoroutineConstructorAdvice.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+public abstract class CoroutineConstructorAdvice {
+    public static final ByteBuddy byteBuddy = new ByteBuddy();
+    public static final Object continuationTracingWrapperClassMonitor = new Object();
+    public static final Object tracingContinuationInterceptorClassMonitor = new Object();
+
+    private CoroutineConstructorAdvice() {
+    }
+
+    @Advice.OnMethodEnter(suppress = Throwable.class)
```
Comment (patiramyadav) [inline apm-agent-plugins/apm-kotlin-concurrent-plugin/src/main/java/co/elastic/apm/agent/kotlinconcurrent/CoroutineConstructorAdvice.java:46]: Here I think we also need to include ```inline=false``` like: ```@Advice.OnMethodEnter(suppress = Throwable.class, inline = false)```

**Comment by jerryjpark** [comment]
Is there anything that needs to be configured like JVM configs for this AOP code to work? I'm getting below error when I launch the application from CLI. It is currently working for me only when I run the application in IntelliJ IDEA.

`java.lang.IllegalStateException: Error invoking (accessor)::defineClass
	at co.elastic.apm.agent.shaded.bytebuddy.dynamic.loading.ClassInjector$UsingReflection$Dispatcher$UsingUnsafeInjection.defineClass(ClassInjector.java:1024)
	at co.elastic.apm.agent.shaded.bytebuddy.dynamic.loading.ClassInjector$UsingReflection.injectRaw(ClassInjector.java:276)
	at co.elastic.apm.agent.shaded.bytebuddy.dynamic.loading.ClassInjector$AbstractBase.inject(ClassInjector.java:113)
	at co.elastic.apm.agent.shaded.bytebuddy.dynamic.loading.ClassLoadingStrategy$Default$InjectionDispatcher.load(ClassLoadingStrategy.java:233)
	at co.elastic.apm.agent.shaded.bytebuddy.dynamic.loading.ClassLoadingStrategy$Default.load(ClassLoadingStrategy.java:144)
	at co.elastic.apm.agent.shaded.bytebuddy.dynamic.TypeResolutionStrategy$Passive.initialize(TypeResolutionStrategy.java:100)
	at co.elastic.apm.agent.shaded.bytebuddy.dynamic.DynamicType$Default$Unloaded.load(DynamicType.java:6154)
	at kotlinx.coroutines.AbstractCoroutine.<init>(AbstractCoroutine.kt:35)
	at kotlinx.coroutines.AbstractCoroutine.<init>(AbstractCoroutine.kt:41)
	at kotlinx.coroutines.future.CompletableFutureCoroutine.<init>(Future.kt:51)
	at kotlinx.coroutines.future.FutureKt.future(Future.kt:42)
`

**Comment by bastrich** [comment]
@jerryjpark how do you build and add an elastic apm dependency? which kotlin version?

**Comment by jerryjpark** [comment]
@bastrich thanks for your reply. I don't have access to my computer now and I won't for couple more days.

I should've shared with you more of the stack trace. The underlying reason (which may or may not be apparent in the stack trace I shared above) is because the Coroutine class the complaining classloader was looking for was not available to the classloader. So, yes it could have to do with the particular version of the class binary the claasloader was looking for. So, I matched the (provided scope) Kotlin dependencies for your plugin to the version of Kotlin used in our project (1.5.0) and rebuilt the plugin to rule out version mismatch as an issue. Still, I cannot say for sure that it is not a version mismatch issue until I debug the Gradle package part. 

We are using Gradle and Gradle plugin to package our application with all its dependencies. We are clearly providing the correct version of the Kotlin dependencies that your plugin expects (at least per the Gradle project def).

Once I return to work, I plan to inspect the fat jar file to begin with to make sure the runtime Kotlin dependencies make to where they should be. Then, I will debug the class loaders if there can an issue for the Kotlin dependency to be found by the relevant classloader that throws CNFE.

**Comment by samuelbirocchi** [comment]
Is this dead?

**Comment by Reizinixc** [comment]
Does anyone have a plan or picked this up recently? It seems the PR is out-of-date for a while.

My team would like to use this to implement a new service instead of doing a workaround like the existing one.

Although we're absolute novices, we may consider making some changes to this.

Currently, I successfully converted the code that uses reflection to non-reflection, but it's very limited and still breaks the app for some cases, e.g.,
* Only `-javaagent` works, `apm-agent-attach` fails miserably to find `kotlinx.coroutines.AbstractCoroutine` (not sure yet why)
* Using a `suspend` function without `async` (need to add code to support thread switching). For instance,
   ```kotlin
   coroutineScope {
       val span = ElasticApm.currentSpan().startSpan("db", "mariadb", "query")

       span.setName("SELECT subdivisions").activate().use {
           fun dashIfEmpty(s: String) = if (s == "") "-" else s

           println("1 - ${ElasticApm.currentTransaction().id.let(::dashIfEmpty)}: ${ElasticApm.currentSpan().id.let(::dashIfEmpty)} (${Thread.currentThread().name})")
           delay(300L)
           println("1A - ${ElasticApm.currentTransaction().id.let(::dashIfEmpty)}: ${ElasticApm.currentSpan().id.let(::dashIfEmpty)} (${Thread.currentThread().name})")
           span.end()
       }
   }
   ```
   ```
   1 - c99bfe8ce5f05d85: a57e817c550748c9 (http-nio-8080-exec-1)
   2022-11-14 09:57:06,835 [http-nio-8080-exec-1] WARN  co.elastic.apm.agent.impl.ActiveStack - Deactivating a context ('ExampleController#controller' 00-b7beea814b3b854c500e814a55408e1e-c99bfe8ce5f05d85-01 (595962ca)) which is not the currently active one ('SELECT subdivisions' 00-b7beea814b3b854c500e814a55408e1e-a57e817c550748c9-01 (4a06cfe3)). This can happen when not properly deactivating a previous span or context.
   1A - -: - (kotlinx.coroutines.DefaultExecutor)
   ```

**Pull request closed without merging**